modules/spotxBidAdapter.js

```diff
@@ -0,0 +1,417 @@
+
+      if (utils.getBidIdParameter('pre_market_bids', bid.params) != '' && utils.isArray(utils.getBidIdParameter('pre_market_bids', bid.params))) {
+        const preMarketBids = utils.getBidIdParameter('pre_market_bids', bid.params);
+        ext.pre_market_bids = [];
+        for (let i in preMarketBids) {
+          const preMarketBid = preMarketBids[i];
+          let vastStr = '';
+          if (preMarketBid['vast_url']) {
+            vastStr = '<?xml version="1.0" encoding="utf-8"?><VAST version="2.0"><Ad><Wrapper><VASTAdTagURI>' + preMarketBid['vast_url'] + '</VASTAdTagURI></Wrapper></Ad></VAST>';
+          } else if (preMarketBid['vast_string']) {
+            vastStr = preMarketBid['vast_string'];
+          }
+          ext.pre_market_bids.push({
+            id: preMarketBid['deal_id'],
+            seatbid: [{
+              bid: [{
+                impid: Date.now(),
+                dealid: preMarketBid['deal_id'],
+                price: preMarketBid['price'],
+                adm: vastStr
+              }]
+            }],
+            cur: preMarketBid['currency'],
+            ext: {
+              event_log: [{}]
+            }
+          });
+        }
+      }
+
+      const mimes = utils.getBidIdParameter('mimes', bid.params) || ['application/javascript', 'video/mp4', 'video/webm'];
+
+      const spotxReq = {
+        id: bid.bidId,
+        secure: secure,
+        video: {
+          w: contentWidth,
+          h: contentHeight,
+          ext: ext,
+          mimes: mimes
+        }
+      };
+
+      if (utils.getBidIdParameter('price_floor', bid.params) != '') {
+        spotxReq.bidfloor = utils.getBidIdParameter('price_floor', bid.params);
+      }
+
+      if (utils.getBidIdParameter('start_delay', bid.params) != '') {
+        spotxReq.video.startdelay = 0 + Boolean(utils.getBidIdParameter('start_delay', bid.params));
+      }
+
+      if (bid.crumbs && bid.crumbs.pubcid) {
+        pubcid = bid.crumbs.pubcid;
+      }
+
+      const language = navigator.language ? 'language' : 'userLanguage';
+      const device = {
+        h: screen.height,
+        w: screen.width,
+        dnt: utils.getDNT() ? 1 : 0,
+        language: navigator[language].split('-')[0],
+        make: navigator.vendor ? navigator.vendor : '',
+        ua: navigator.userAgent
+      };
+
+      const requestPayload = {
+        id: channelId,
+        imp: spotxReq,
+        site: {
+          id: siteId,
+          page: page,
+          content: 'content',
+        },
+        device: device,
+        ext: {
+          wrap_response: 1
+        }
+      };
+
+      if (utils.getBidIdParameter('number_of_ads', bid.params)) {
+        requestPayload['ext']['number_of_ads'] = utils.getBidIdParameter('number_of_ads', bid.params);
+      }
+
+      const userExt = {};
+
+      // Add GDPR flag and consent string
+      if (bidderRequest && bidderRequest.gdprConsent) {
+        userExt.consent = bidderRequest.gdprConsent.consentString;
```

Comment:
If a pub enables the consent modul, but for some reason the consent information cannot be found,
then prebid passes an object like so to the bidders:

```javascript
{
   gdprConsent: {
      consentString: undefined
      gdprApplies: undefined
      vendorData: undefined
  }
}
```

So technically your `userExt` would be:

```javascript
{
   consent: undefined
}
```

at this point, but it gets pruned before the request eventually gets sent by prebid so no big deal, just though I would make aware that Prebid passes those undefined values in some instances!

Reply:
Our ad server should handle undefined in those requests anyways, but that's good to know!